File: apps/backends/node/src/handlers/checkout.ts
import { FastifyRequest, FastifyReply } from 'fastify';
import { Pool } from 'pg';
import { processCheckout } from '../services/checkout.js';
import { CheckoutRequest } from '../types.js';

export async function checkoutHandler(
  req: FastifyRequest,
  reply: FastifyReply,
  pool: Pool,
  taxServiceUrl: string
) {
  try {
    const result = await processCheckout(req.body as CheckoutRequest, pool, taxServiceUrl);
    reply.status(201).send(result);
  } catch (error: any) {
    const msg = error.message || 'Internal server error';
    if (msg.includes('Invalid')) {
      reply.status(400).send({ error: msg });
    } else if (msg.includes('stock') || msg.includes('Cart must')) {
      reply.status(422).send({ error: msg });
    } else {
      reply.status(500).send({ error: msg });
    }
  }
}

File: apps/backends/python/app/handlers/checkout.py
from fastapi import HTTPException
import asyncpg
from app.models.types import CheckoutRequest, CheckoutResponse
from app.services.checkout import process_checkout

async def checkout(
    req: CheckoutRequest,
    pool: asyncpg.Pool,
    tax_service_url: str,
) -> CheckoutResponse:
    try:
        return await process_checkout(req, pool, tax_service_url)
    except ValueError as e:
        msg = str(e)
        if 'Invalid' in msg:
            raise HTTPException(status_code=400, detail=msg)
        elif 'stock' in msg or 'Cart must' in msg:
            raise HTTPException(status_code=422, detail=msg)
        else:
            raise HTTPException(status_code=500, detail=msg)
